mne/io/reference.py

```diff
@@ -467,39 +476,53 @@ def set_bipolar_reference(inst, anode, cathode, ch_name=None, ch_info=None,
         raise ValueError('Number of channel info dictionaries must equal the '
                          'number of anodes/cathodes.')
 
-    # Merge specified and anode channel information dictionaries
-    new_chs = []
-    for ci, (an, ch) in enumerate(zip(anode, ch_info)):
-        _check_ch_keys(ch, ci, name='ch_info', check_min=False)
-        an_idx = inst.ch_names.index(an)
-        this_chs = deepcopy(inst.info['chs'][an_idx])
-
-        # Set channel location and coil type
-        this_chs['loc'] = np.zeros(12)
-        this_chs['coil_type'] = FIFF.FIFFV_COIL_EEG_BIPOLAR
-
-        this_chs.update(ch)
-        new_chs.append(this_chs)
-
     if copy:
         inst = inst.copy()
 
-    for i, (an, ca, name, chs) in enumerate(
-            zip(anode, cathode, ch_name, new_chs)):
-        if an in anode[i + 1:] or an in cathode[i + 1:] or not drop_refs:
-            # Make a copy of the channel if it's still needed later
-            # otherwise it's modified inplace
-            _copy_channel(inst, an, 'TMP')
-            an = 'TMP'
-        _apply_reference(inst, [ca], [an])  # ensures preloaded
-        an_idx = inst.ch_names.index(an)
-        inst.info['chs'][an_idx] = chs
-        inst.info['chs'][an_idx]['ch_name'] = name
-        logger.info('Bipolar channel added as "%s".' % name)
-        inst.info._update_redundant()
+    anode = _check_before_reference(inst, ref_from=cathode,
+                                    ref_to=anode, ch_type='auto')
+
+    # Create bipolar reference channels by multiplying the data (channels x
+    # time) with a matrix n_virtual_channels x channels
+    # and add them to the instance
+    multiplier = np.zeros((len(anode), len(inst.ch_names)))
+    for idx, (a, c) in enumerate(zip(anode, cathode)):
+        multiplier[idx, inst.ch_names.index(a)] = 1
+        multiplier[idx, inst.ch_names.index(c)] = -1
+
+    # Create new instances (create each from single channel to allow same
+    # channel in anode multiple times)
+    ref_instances = list()
+    for ch_idx, (an, ca, name, info) in enumerate(zip(anode, cathode,
+                                                      ch_name, ch_info)):
+        _check_ch_keys(info, ch_idx, name='ch_info', check_min=False)
+        ca_idx = inst.ch_names.index(ca)
+
+        # Merge specified and anode channel information dictionaries
+        add_inst = inst.copy().pick(an)
+        add_inst.rename_channels({an: name})
+        # Use location of cathode
+        add_inst.info['chs'][0]['loc'] = inst.info['chs'][ca_idx][
+            'loc'].copy()
+        add_inst.info['chs'][0]['coil_type'] = FIFF.FIFFV_COIL_EEG_BIPOLAR
+        add_inst.info['chs'][0].update(info)
+        ref_instances.append(add_inst)
+
+    ref_inst = ref_instances[0].add_channels(ref_instances[1:],
+                                             force_update_info=True)
+
+    if isinstance(inst, BaseEpochs):
+        ref_inst._data = np.asarray([multiplier.dot(ep) for ep in inst._data])
```

Comment:
I would leave this for the end but there might be some performance benefits of using [einsum](https://numpy.org/doc/stable/reference/generated/numpy.einsum.html). I always get confused when using it though

Reply:
`@` has much saner broadcasting behavior than `np.dot` (`@` operates along the last two dims of both inputs, and np.dot... doesn't), this is probably the same as `multiplier @ inst._data`

Reply:
I came across einsum or [block-matrices](https://numpy.org/doc/stable/reference/generated/numpy.block.html) too when searching for a more elegant solution. (EDIT: But I couldn't solve it yet)
It appears to me, that the rereferencing itself doesn't take as much of the computation time as the overhead from creating and checking info in add_channels. You can try it out, if you replace l. 509-511 with:
```python
    if isinstance(ref_instances, list):
        ref_instances = add_inst
    else:
        ref_instances.add_channels([add_inst], force_update_info=True)
ref_inst = ref_instances
```